### ISSUE TITLE
StepMania 6 super mega ultra planning issue

### ISSUE THREAD

**Issue by shakesoda** [stepmania/stepmania#740]
Just so we've got one place with public visibility that isn't an outdated wiki page.

For context, StepMania is a mess. We know it's a mess, it's been a mess since... well, always. We've got insane decade-old bugs which we can't even begin to comprehend and the original maintainers are long gone. We've brought it up regularly for years that this situation can't last forever, while we kept working on making 5.0 all that it could be. We've even got 5.1 in the works right now to break what needs breaking and improve near-term things which can be fixed. But we'll need more than that, and with a stable release out the door we're in a good position to think and start acting on this.

Tentative plan, which I've already started on locally:
- Fork [LOVE 0.11](https://bitbucket.org/rude/love) to use as a base. This isn't just because I'm some kind of fanatic, it has real benefits which align with our existing plans anyhow:
  - Totally replaces enormous amounts of decade-old Rage code which nobody even understands
  - Kills lots of kludges for ancient hardware like Voodoo 3's
  - Modern rendering code, including...
  - Real Android support
  - A Raspberry Pi port with total feature parity (performance is kinda TBD, though)
  - iOS support (although we may not maintain this)
  - LuaJIT
  - Code which has been _actually designed_ and not just cobbled together on an as-needed basis.
  - Uses SDL2, as we've been planning to bring in for a good while to fix a whole ton of platform problems.
  - Gamma correction (this significantly improves the quality of blending + text will look better if enabled)
  - Font rendering that's actually good (ttf/otf support, bmfont support, kerning)
  - Unicode text input
  - Proper mouse/touch support
  - Sane threading in Lua
  - Networking in Lua (via TCP/UDP sockets and enet)
  - Arbitrary shaders without rebuilding RageDisplay
  - The list goes on... you should get the idea by now. It offers a lot that we could use.
- Trim the fat and convert things to SM's code style
  - This also means formally defining SM's code style, and we can use something more sane than 8-width tabs. This will be specified in another issue, _do not discuss here_.
  - We won't need to run game archives like Love does, but the same system can be used to reimplement SMZIP support.
  - Love has some modules we won't be needing, like Box2D.
  - We'll also need to add a third dimension to a bunch of things, but that isn't a huge undertaking.
- Bind and refactor useful things in the existing codebase. Obviously, there's some stuff we've already got which is better not being rewritten or is still fresh. Some examples:
  - @kyzentun's new notefield
  - The note loaders (tentatively, at least. I'd rather these be in Lua)
  - The threaded light/input drivers
  - Movie decoders, maybe. I'd honestly like to nuke ffmpeg, but threaded video decode is a must.
    - Love actually has this, but currently only for Theora.
  - Crypto stuff
  - TimingData and friends. This stuff could be rebuilt in Lua, but that's a major project in itself.
- Overhaul the UI using our new superpowers
  - Proper touch/mouse support on everything.
  - Menus that aren't garbage - we can clean them up a lot just by not forcing every inappropriate thing into ScreenSelect...
  - Kill metrics.ini once and for all (but you could by all means have one for your theme, if that's your thing, but we can kill the internals using it.
  - Potentially convert language files to a standardized format (.mo/po files)
- Include SQLite bindings. I've already written them and it's really, really convenient to have.
- Overhaul the caching system since we won't have things which depend on the old behavior anymore
  - Background loading of as much as possible, so the game starts up immediately.
  - Binary cache is worth investigating
- User-defined game types and styles.
- Actually write tests for new code and refactor anything we can to be testable.

**Before ripping my head off over proposing an effective rewrite and saying that this will take forever - this is a _long term_ plan.**

I've also been accused of using the Love-hammer on every nail, that's not what this gets down to. It brings a lot to the table for our severely unmaintainable codebase. A big solution is needed for a big problem and we've known this for years.

It goes without saying that I'm willing to put my code where my mouth is on this, but it should also go without saying that I want as many people on board as possible.

Let's hear those complaints.

Suggestions gathered from the comments:
- Look for new simfiles at all times in a background thread, cache/uncache as needed.
- Lirodon proposed a package schema, seems reasonable to me. https://gist.github.com/ListenerJubatus/e77fa67e2ebaa461e0b7
- Kill the banner cache, we can just lazy load images in another thread when they're needed.

**Comment by cybik** [comment]
"This is a bad idea shame on you". <3

Depending on the necessary core re-engineering, this is either genius, or crazy.

**Comment by KcHecKa** [comment]
Maybe you should also fix the windowed/full screen thingy. Like, make edge
less window full screen. Like how osu is.

**Comment by shakesoda** [comment]
@KcHecKa Yeah, that would be something we get without any extra work. I've wanted that fixed for ages, too, exclusive fullscreen sucks!

**Comment by cybik** [comment]
Admittedly my Android branch is still, well, kind of broken.

**Comment by ListenerJubatus** [comment]
I still think we should adopt some sort of schema-based package system like I proposed in https://gist.github.com/ListenerJubatus/e77fa67e2ebaa461e0b7

**Comment by AldoMX** [comment]
TL;DR illustrated version:
![image](https://cloud.githubusercontent.com/assets/476531/9209023/0b82f2b4-403a-11e5-8044-d707b463284e.png)

**Comment by KcHecKa** [comment]
Nice!! Lol

**Comment by wolfman2000** [comment]
I remember trying to experiment with SQLite before. I don't recall why that was abandoned.

I can look up Love later this week. Slightly busy for the next few days.

**Comment by SpoOkyMagician** [comment]
You got me at this:

"Kill metrics.ini once and for all (but you could by all means have one for your theme, if that's your thing, but we can kill the internals using it."

+1

**Comment by YungDaVinci** [comment]
Sounds great, although how exactly are user defined game types gonna work?
And packages would be pretty dope too.

**Comment by Kingdaro** [comment]
Even if this never happens, I'm so happy things are at least hopefully possibly moving in a better direction.

**Comment by Xaero252** [comment]
Dynamic cache updating and loading. This way someone managing a number of installs remotely could "push" song directories to the installs and have the songs show up in the menu without having to reload the whole engine. It's 2015, we have the cpu power and memory to handle this.

**Comment by kyzentun** [comment]
@Xaero252 I spent a lot of time on implementing more advanced song sorting, and as a result, practically nobody uses Consensual besides me.  So I feel like better song management is a complete waste of dev time unless the dev implementing it wants it for their personal use.

**Comment by Xaero252** [comment]
If the cache system is getting an overhaul anyways I don't think this would
be too much of a misdirection.  I don't care about song sorting.
This would be an engine feature not a theme feature (I.e. it would affect
caching and loading, not sorting options)

**Comment by kyzentun** [comment]
That's exactly my point.  Song sorting makes a big difference to me, but 99.9% of other players don't care because they only hae 500 songs.
Dynamic song loading makes a big difference to you, but most people don't care because restarting stepmania takes under a minute.
(if restarting takes more than a minute, go into Preferences.ini and set FastLoad=1 and BannerCache=Off  Then it'll stop looking inside folder of songs that are already cached, which is a waste of time)

**Comment by Xaero252** [comment]
So, by proxy - we should avoid making performance enhancements like making
menu loading multi-threaded and cached - since only a handful of people
actually give a crap about the performance of SM and the rest don't mind
unless it becomes unusable?
I mean, where do we draw the line on implementing features and
enhancements?
I understand in the past 7 years or so I have made basically no
contributions to the code, so my opinion matter little but still.

**Comment by cybik** [comment]
I'll settle this discussion between Xaero and Kyz with one simple sentence:

Pull Requests / Code Reviews are welcome :)

I would personally back that feature; I find it rather interesting, and on low-powered and/or "removable storage" systems, it kind of makes sense.

Consider this: a simple arcade where one plugs in one's USB stick with profiles AND songs, and uses the arcade to play their songs. Boom. Use case dropped.

**Comment by shakesoda** [comment]
I don't see much of a reason we couldn't watch the song directories for additions/deletions. Doesn't osu do that too?

I'm throwing it in the list.

**Comment by TonyTheTGR** [comment]
Well, design begins with intent. SO! What interface do we WANT? What features do we want? Where is the "minimum viable project," and the finished, "feature-rich project?"

**Comment by TonyTheTGR** [comment]
Minimum:

Title Screen (Play, Reload Song Data, Controls, Exit)
Control Config Screen
Style Select Phase
Song Select Phase
Gameplay Phase
Evaluation Phase
Game Over/Fail
Game Over/Win

Feature Rich:
Modifiers Phase
System Options
Course Select Phase
Course Creation Phase
Edit Mode

Netplay Mode (2-5 players: player is centered, other network players in surrounding areas; uses host's songlist and host controls song selection, but other players can adjust difficulty/modifiers, etc. Only your timing must be precisely measured; other players' results simply copypaste across - usually useful for sharing edits without copying SM files). Can play versus on each machine, for two scores (and up to 10 players? or 5 Doubles?)

Master Mode via grouping/difficulty indexing (1-5 ft, 6-10 ft, 10-14 ft, 15-n ft)
Endless Mode (with reasonable repeat-exclusion buffer, like a song can't reappear within ten to fifty stages, depending on total tracks in the group)
6-Panel generation script that doesn't suck (I have an idea that may, uses Double charts though)
"Keyword Course" design - a crs of keywords that looks for songs featuring those words?
RDA/RDI-style Extra Stage system
Unlock System(s)

*shrug*

**Comment by TonyTheTGR** [comment]
Do we want local 4-player, or Triple/Quad Mode? Stuff like Attack Mode from Ultramix, or that auto-medley thing Universe 3 does? I mean, I know how to math that stuff; but that doesn't mean I fully know how to *program* it.


New(Stepchart) =

Song1.chart + Song2.chart + Song3.chart + Song4.chart...
(if fewer than max stages, it'll just copypaste "nothing" anyways?)

Dim BPMBuffer as (BPM.Song1.chart.max - 32) - (BPM.Song2.chart.1)

At Measure(Song1.chart.max - 32)
Play Song1.MusicSpeed
And Song2.MusicSpeed - BPMBuffer

At Measure(Song1.chart.max - 24)
Play Song1.MusicSpeed + BPMBuffer * 0.75
And Song2.MusicSpeed - BPMBuffer * 0.25

At Measure(Song1.chart.max - 16)
Play Song1.MusicSpeed + BPMBuffer * 0.5
And Song2.MusicSpeed - BPMBuffer * 0.5

At Measure(Song1.chart.max - 8)
Play Song1.MusicSpeed + BPMBuffer * 0.25
And Song2.MusicSpeed - BPMBuffer * 0.75

At Measure(Song1.chart.max)
Play Song2.MusicSpeed * 1.0
And Fade(Song1.MusicSpeed - BPMBuffer)

Dim Song2.chart.min += Song1.chart.max
Dim Song2.chart.max += Song1.chart.max

...I hope it makes sense where I'm going with this. It's not actual code necessarily, but is it followable?

**Comment by SoonDead** [comment]
Let me do a little necromancy.

@TonyTheTGR mentioned, local 4 player support. This is something that has been hacked around countless times in forks. I would very much like to see a good solution for this.

The 2 main solutions I have seen for this is:

 1, Using a stepmania fork that is capable of handling 4 players on one screen.

For example [this 3.9 fork](https://cspotcode.wordpress.com/4-player-stepmania/).

Problems with forks is, they tend to be left behind. Never updated and never merged back. Especially hacky stuff like this.

Also 4 players don't properly fit on 1 screen in a lot of cases, might need multi-display support which is a pain.

 2. Using a lan synchronization between 2 identical machines.

The problem with this is the inaccurate sync between 2 machines. A 0.5 sec sound difference between the 2 is especially annoying if the machines are next to each other.

Also this needs 2 machines.

<br>

It would be great if SM 6 had proper support for either of these cases.

I have requested this on the [old bug tracker](http://ssc.ajworld.net/sm-ssc/bugtracker/view.php?id=640).

**Comment by kyzentun** [comment]
Anybody who implements 4 player support and has to break some existing content like themes or even poorly coded gimmick simfiles will face strong opposition from the community.  So it's hardly worth considering doing for SM5 or 5.1.
(for example: dbk2 got sent as a mediator from certain gimmick authors complaining about 5.1 using a new mod system that will work better with more players than the existing one)

SM6 will be a complete rewrite, so using an ancient fork from 3.9 wouldn't be an option even if it was a good idea.

**Comment by shakesoda** [comment]
There's no reason we can't support >2 players in 6, it'd be nice to have. It's mostly a design problem for the interface.

LAN sync I'd consider part of rebuilding SMO, although matching audio timing might have to just be ignored (best you can do is make sure your global offsets are correct - the packet delay on LAN is generally negligible).

**Comment by SoonDead** [comment]
>Anybody who implements 4 player support and has to break some existing content like themes or even poorly coded gimmick simfiles will face strong opposition from the community. So it's hardly worth considering doing for SM5 or 5.1.

Whaat? Isn't this an "SM6 rewrite" issue? Correct me if I'm wrong, but in a full rewrite gimmick simfiles will break anyway.

About not breaking existing themes:

It could be part of a theme how many players it supports. The core engine could support as many as possible, but depending on the theme there might be less players available.

Legacy themes (if you want to even support legacy themes in SM6) where available player count is not defined, could be treated as "1 or 2 players" themes. This way they will not break in a UI perspective.

Specifiable player count support would also lessen the burden on new theme authors as they would not have to support a 64 player extravaganza if they decide not to.

>SM6 will be a complete rewrite, so using an ancient fork from 3.9 wouldn't be an option even if it was a good idea.

I sure hope noone wants to merge that into SM6 :D (no offense to cspotcode, it's quite a feat what he did in this fork).

>LAN sync I'd consider part of rebuilding SMO, although matching audio timing might have to just be ignored (best you can do is make sure your global offsets are correct - the packet delay on LAN is generally negligible).

If the delay between the 2 machines is stable and the user can correct it with an offset it sounds perfect.

A "sync machines" screen would be very good for this, where 2 (or more) machines can connect to each other, play the same music at the same time and the offset can be adjusted on each one on the fly. This way it would be pretty intuitive and fast to set this offset up perfectly.

**Comment by Flameshadowxeroshin** [comment]
One thing I'd really like to see is most of the note judgment stuff out in Lua, or at least able to be out in Lua.

**Comment by xxsodapopxx5** [comment]
I would like to point out that there is not a lot of information on aiding in development for Stepmania. I am a c# developer and would love to learn how to get into the development for this project, but there is just a lack of information(or i am naive and can't find it) on how to get started into aiding in development for this. 

You have the 'Compiling' page http://www.stepmania.com/wiki/compiling-stepmania/ , but it is lacking. I think a lot of code cleanup process would happily be handled by developers if they knew the best way to get started into this project. A Code cleanup Milestone with a task list style breakdown would be great as well, that way people aren't duplicating work.

The more developers that are interested in the project the faster features and code cleanup can be written!

**Comment by Flameshadowxeroshin** [comment]
StepMania 6 won't use the same codebase as StepMania 5, and it is true that there's no introduction to the development process, mainly because it's basically completely undirected right now and the SM6 complete rewrite is supposed to change that, as far as I can tell.

**Comment by kyzentun** [comment]
I got started in this project by thinking of things I wanted and figuring out how to do them.  Start small and work your way up as you gain experience is the best advice for joining that I can give.

**Comment by monokles** [comment]
I hope this will allow for a proper rewrite of the online code as well; from the looks of it LOVE has some very easy to use networking libraries.

**Comment by Lebon14** [comment]
And while we're at it, why not add more music formats other than MP3 and OGG like AAC (m4a), FLAC, etc... FLAC shouldn't even be an issue since the libraries are open source and you load and unload the song into memory only when needed.

**Comment by kyzentun** [comment]
Sure, go ahead, write the code and submit a pull request.  The rest of us have other things to work on.

**Comment by shakesoda** [comment]
It's pretty much useless to support tons of audio formats when all the content needs to be tailored for the game regardless, and seeking with various libs isn't always good (it even sucks now with mp3).

That said, I think this plan has a pretty fatal flaw: nobody wants to implement something this big. I'd like to think about what can be done to make it more incremental.

**Comment by Prcuvu** [comment]
Maybe we can make FFmpeg do audio decoding as well as video decoding.

**Comment by Lebon14** [comment]
You guys have a feature that decodes video and that's pretty hard to do with the tons of containers, codecs, etc around. If there's such a feature, why not more audio format than MP3 and OGG? The libraries for video decoding should also have libraries for audio too since they go hand in hand. In fact, FLAC is supported by ffmpeg. There's no excuse.

**Comment by shakesoda** [comment]
Most audio formats just aren't worth the time, including FLAC, and ffmpeg is one of the biggest pains in the asses possible. If so many files didn't depend on it working with all kinds of random crap I'd remove it and just support webm and ogg.

**Comment by Xaero252** [comment]
Yeah ffmpeg is a literal nightmare to work with. It's fine for an end-user
but for developers its basically in a constantly shifting state. This is
why older versions of SM were compiled against a static FFMPEG version. You
couldn't have a single ffmpeg revision worth of changes without everything
breaking. As far as supporting other audio formats - most of them aren't
worth supporting. It'd be nice to support FLAC, but the problem is that
it's already common practice within the community to use lower quality
MP3's or OGG's to shrink the overall file size of simfile packs; Not only
that but just like supporting vectors for theme elements (which would be
awesome, you could have infinitely scaling aspect-ratio independent themes)
there's a certain hardware overhead and latency associated with the format.
It takes too long to render, just like vector graphics.

Note that I don't contribute much to SM; but these are just general
observations to be made about supporting such technologies in a realtime
application. It's just not viable.

**Comment by SpoOkyMagician** [comment]
Honestly, I don't see why you even bother with mp3 support. MP3 VBR's screws sync up anyway... If I was to recommend how to start, I would just stick with ogg first. (and webm. I dunno much about that.) Then, work your way up from there. Do what you know and do it well.

**Comment by TonyTheTGR** [comment]
Well, tbh; mp3 came first. It was the major, common format used in similes
since forever. 4.0 added ogg vorbis *specifically* for ITG. MPEG and AVI so
far are strictly for BG videos... though they can contain audio and be
heard, we used to need them separate so in case of VRAM failure, we could
cut the video without cutting sound (the crucial game element, versus the
cosmetic ones).

Bear in mind, 3.9 was common on XP machines, and we had much tighter
constraints then than we do now.

**Comment by shakesoda** [comment]
https://github.com/stepmania/six

not playable, but work has started. this issue tracker/wiki/etc will still be used for it and in the future it will be merged into a branch of this repo.

**Comment by karai17** [comment]
Just a reminder for Issue #5, it would be really handy to be able to "star" a sing and have it show up in a "Favourites" category so that having 99999 songs isn't quite as burdensome as it normally is. Combining each active profile's favourites into a single category would be useful too, instead of only using P1's.

Also using a hashing algorithm to detect if each player has the right version of a song for netplay. I've noted several instances where me and a friend had two very different versions of the same song and netplay could not detect this. A simple md5 hash should be enough when caching the song. Not sure what exactly should be hashed, though (sm file? whole directory?).

**Comment by karai17** [comment]
Another interesting feature for netplay would be the ability to share songs. This would be hugely useful when playing with random people online since it's less likely you'll be able to communicate outside of SM to share content. The host could see which songs all players in the room have, and which songs only the host has and offer to share a song if not everyone has it. If everyone accepts, the song is sent out. If it is declined, it is cancelled. Users who receive the song can store it in a temp location and then after the round, are given an option to keep the song (sent to a "From Netplay" directory, or maybe if there is package info, sent to a proper directory)

**Comment by fent** [comment]
Identifying a song based on its metadata and not its path would be nice. Sorting songs by name would then not result in having as many copies of the same song with the same steps.

Would also facilitate keeping track of scores. Some people look for a song through its original pack, others look at a collection pack from a tournament, and others order by name in which they'd usually pick whichever.

**Comment by Xaero252** [comment]
The problem with using metadata is that historically, almost no songs have
present, intact, correct or proper metadata. That said I see your point
about duplicate songs, and have a solution. While calculating the md5 sum
of the songs themselves would be problematic; calculating the sum of JUST
the step data from the stepchart should be a negligible resource expense
and allow one to "negate" duplicate entries. There would still be edge
cases unaccounted for (people adding a single chart to a simfile, for
example would break this) but for the majority of cases it would eliminate
duplicate entries in the song wheel. In this manner you would store the
title of the song and the MD5 sum as a pair for cache purposes. Any entries
with different values would obviously be unique entries and could be
treated as such.
Seems logical. Avoids relying on mostly non-existent metadata. Doesn't add
new requirements to *.sm files for implementation. This would of course
make caching take longer - perhaps an optional feature or something that
gets ignored when FastLoad=1.

**Comment by karai17** [comment]
One major problem with doing *anything* would be breaking large swaths of legacy content. Upgrading form SM3.9 to SM5 breaks all themes and noteskins, for instance. It is definitely not ideal to break content.. But you can't always be held back by legacy, either. I don't think it is the best solution to make SM6 something 100% incompatible with all current content, but it is definitely worth thinking about what we can do to make legacy content still function while making future content better.

**Comment by karai17** [comment]
We could start deprecating content and give people time to update content that is still valued before removing support for legacy content. We could also make tools that help convert legacy content to the newly supported format.

**Comment by Xaero252** [comment]
This is true. More advanced sorting options, for example - while mentioned
previously in this discussion as being "greatly undervalued" by the
community, might be more well received if it was supported by a platform
people could get behind. For example, the majority of the arcade use
community was stuck on OpenITG/3.95 based builds due to their compatibility
with a particular set of hardware and software requirements. SM5 brought a
lot to the table and many people were reluctant to move to it - but now
that SM5 has begun to meet feature parity with OpenITG in those areas,
people are more inclined to switch to the new platform. I think the same
could be said about implementing more advanced sorting options based on
music genre, chart density and other such metrics - including the ability
to weed out duplicate charts. As long as the platform can be widely
received, the new features would be more prominently available to the end
user.
Food for thought. I agree that widely StepMania has a lot of stagnated
legacy support. How much of what was lost in the wake of changes between
3.9 and 5 was content worth keeping? And of that - how much wasn't just
ported to the new platform? Not only that, but this is an opportunity to
make content creation more accessible. The old theme metrics and legacy
NoteSkin/Announcer/Dancer implementations were somewhat convoluted and
difficult for people to understand. Plus there wasn't really a standardized
tool set for working with these things.

**Comment by kyzentun** [comment]
This is probably my current favorite thread on more advanced sorting options:  https://github.com/stepmania/stepmania/issues/1288

**Comment by karai17** [comment]
Yeah, exactly. If we give people a time buffer and say "all content is gonna be broken in 6 months, but here is how you can **easily** port your content from SM5 to SM6", I think it would give a much better message that we *want* content to work, instead of "this is how it's gonna be, deal with it". The rhythm community is well known for having strong egos, so being as inviting as possible would be a good way to coax people into buying in to the new systems.

**Comment by SoonDead** [comment]
@karai17 I would love to send my `petya_ransomware.ogg` to everyone playing against me :)

In all seriousness, song sharing would require very strict file restrictions.

**Comment by AldoMX** [comment]
I heard you didn't want to struggle with FFMpeg, so I'm sharing the `RageSoundReader_FFMpeg` file I wrote for 3.9:
https://gist.github.com/AldoMX/ce64f6fa2d1e50e4289841cf92205a76

It is tested with FFMpeg 2.1.4, which was the latest version when I updated `MovieTexture_FFMpeg` 3 years ago.

For Windows you can get the headers/libs from here:
https://ffmpeg.zeranoe.com/builds/win32/dev/ffmpeg-2.1.4-win32-dev.7z

And the DLLs from here:
https://ffmpeg.zeranoe.com/builds/win32/shared/ffmpeg-2.1.4-win32-shared.7z

Hopefully it will be less tedious for you to port this file from 3.9 to 5.0 than to write a SoundReader from scratch.

**Comment by diamondman** [comment]
@shakesoda I like the idea of using an open 2D engine like LOVE2D for stepmania. There is no reason to maintain the full engine when that is being done in another project. 

@shakesodaIn 2015 you said that you had some progress on porting over to LOVE. What branch is that on? I think there would be a decent number of people who would work on small pieces if it is available and we know what is broken (I was looking to see if I could add code to it).

I also would like to pose a question that I think is important when considering such a drastic rewrite. I see that the requirements for Stepmania are surprisingly low. If we make a rewrite that is trivial to maintain and came with every feature mentioned in this thread (including new stuff), how high are we willing to bump up the system requirements? (For example, if Raspberry Pi 2 is no longer supported and the min becomes a more modern Intel Atom processor or higher performance ARM core)

**Comment by diamondman** [comment]
There also seems to be some complexities with the LOVE build system that would not gel super well with stepmania's entirely CMAKE operations.

Some of the complexities are discussed here https://bitbucket.org/rude/love/pull-requests/79/add-pluggable-modules-to-the-build-system/diff

**Comment by Flameshadowxeroshin** [comment]
@diamondman what progress there is is in the stepmania/six repo.

**Comment by a-shark** [comment]
As long as we're spitballin' about massive Stepmania shakeups/rewrites, in the perfect world I visit in my dreams Stepmania is a libretro core.

https://github.com/libretro/RetroArch

**Comment by Flameshadowxeroshin** [comment]
I have no idea what StepMania being a RetroArch core would gain, and as far as I know it would increase complexity.

**Comment by a-shark** [comment]
It'd remove the need for an input mapping interface, and much of the graphics/audio handling & config. Networking could largely be offloaded to libretro/RA, if that's relevant. Whole point is that it keeps you from having to worry about that kind of common-to-many-gaming-programs stuff and makes portability easier, while also not forcing the user configure their input/graphics/sound over and over for every single thing they want to run. Not sure how it could do anything but _decrease_ complexity, especially when the context of the discussion is already a rewrite/refactor to a different engine.

Plus it'd keep me from needing to run anything other than RetroArch/Lakka on my TV-centered gaming machine. So it'd decrease the complexity of my _use_ of Stepmania, for sure :-)

**Comment by Flameshadowxeroshin** [comment]
An input mapping interface isn't really that complicated, and besides StepMania's input mappings really don't overlap that well with standard RetroArch inputs, so you'd probably just have to reconfigure it again anyway. (But now you can reconfigure it again in the really confusing RetroArch interface! Wow!) We also wouldn't be able to target any version of OGL post 2.0. I don't want rewind and as far as I can tell the best you can do with libretro's builtin networking is couch multiplayer over the internet, which isn't what the goal of online StepMania is anyway. You'd probably have to deal with some libretro backends having crappy audio implementations and people thinking that's a StepMania bug. It pretty much looks like a losing situation all around to me honestly.

**Comment by a-shark** [comment]
I don't get how the input mappings don't overlap well since it's arrows and a handful of action buttons, like everything else (mame works fine, for instance) or how Rarch's mapping interface is any better/worse than the usual (so, fairly blah, but that's the same as all the others, plus it has better auto-mapping than others I've seen, and the ability to save your mappings so they'll auto-map for future devices of the same kind, which is great). But yeah, if you don't think it'll work for a variety of reasons that's fine.

Just throwing my dream out there. Guess a dedicated manually-configured Stepmania-only machine will remain part of my life for the foreseeable future, alongside Lakka/Retroarch and Kodi boxes. Oh well, first world problems and all that.

**Comment by Xaero252** [comment]
This also wouldn't work simply because it goes against the retroarch and
libretro philosophy. It's the same reason they don't want Steam games and
App launchers. If it's not coded from the ground up for their ecosystem it
doesn't belong.
They are pretty set on that.

**Comment by shakesoda** [comment]
I think it's a neat idea but I'm not going to implement StepMania as a RetroArch core.

**Comment by shakesoda** [comment]
after a lot of thought about who doing this is really for and who actually has motivation to implement this (nobody), I don't think this is how SM6 is going to go down. other development will proceed as planned with 5.1 sometime soon and 5.2 later.

I don't think rebuilding the engine to do fundamentally the same thing and still calling it stepmania would actually benefit users as much as just cleaning things up from how they are, and it'd take an order of magnitude more time.
